src/server/debug/server/server.go

```diff
@@ -82,123 +80,59 @@ func NewDebugServer(env serviceenv.ServiceEnv, name string, sidecarClient *clien
-					}
-					return errs
-				case *debug.Filter_Worker:
-					if f.Worker.Redirected {
-						var errs error
-						// Collect self if we're not a worker.
-						if s.sidecarClient == nil {
-							return collect(ctx, tw, pachClient, client.PPSWorkerSidecarContainerName)
-						}
-						// Collect the user container.
-						if err := collect(ctx, tw, pachClient, client.PPSWorkerUserContainerName); err != nil {
-							errors.JoinInto(&errs, errors.Wrap(err, "collect user container"))
-						}
-						// Redirect to the storage container.
-						r, err := redirect(ctx, s.sidecarClient.DebugClient, filter)
-						if err != nil {
-							errors.JoinInto(&errs, errors.Wrap(err, "collect storage container"))
-						}
-						if err := collectDebugStream(tw, r); err != nil {
-							errors.JoinInto(&errs, errors.Wrap(err, "collect debug stream"))
-						}
-						return errs
-					}
-					pod, err := s.env.GetKubeClient().CoreV1().Pods(s.env.Config().Namespace).Get(ctx, f.Worker.Pod, metav1.GetOptions{})
-					if err != nil {
-						return errors.Wrapf(err, "getPod(%s)", f.Worker.Pod)
-					}
-					if err := s.handleWorkerRedirect(ctx, tw, pod, collectWorker, redirect); err != nil {
-						return errors.Wrapf(err, "handleWorkerRedirect(%s)", f.Worker.Pod)
-					}
-					return nil
-				case *debug.Filter_Database:
-					if wantDatabase {
-						return s.collectDatabaseDump(ctx, tw, databasePrefix)
-					}
-				}
-			}
+// the returned taskPath gets deleted by the caller after its contents are streamed, and is the location of any associated error file
+type taskFunc func(ctx context.Context, dir string) error
+type reportProgressFunc func(ctx context.Context, progress, total int)
 
-			// No filter, return everything
-			var errs error
-			if err := collectPachd(ctx, tw, pachClient, pachdContainerPrefix); err != nil {
-				errors.JoinInto(&errs, errors.Wrap(err, "collectPachd"))
-			}
-			pipelineInfos, err := pachClient.ListPipeline(true)
-			if err != nil {
-				errors.JoinInto(&errs, errors.Wrap(err, "listPipelines"))
-			}
-			for _, pipelineInfo := range pipelineInfos {
-				if err := s.handlePipelineRedirect(ctx, tw, pachClient, pipelineInfo, collectPipeline, collectWorker, redirect); err != nil {
-					errors.JoinInto(&errs, errors.Wrapf(err, "handlePipelineRedirect(%s)", pipelineInfo.GetPipeline()))
-				}
-			}
-			if wantAppLogs {
-				// All other pachyderm apps (console, pg-bouncer, etcd, pachw, etc.).
-				if err := s.appLogs(ctx, tw); err != nil {
-					errors.JoinInto(&errs, errors.Wrap(err, "appLogs"))
-				}
-			}
-			if wantDatabase {
-				if err := s.collectDatabaseDump(ctx, tw, databasePrefix); err != nil {
-					errors.JoinInto(&errs, errors.Wrap(err, "collectDatabaseDump"))
-				}
-			}
-			return errs
-		})
-	})
+// TODO:
+// Is it good that DumpRequest contains specific pods? The benefit is explicitness.
+// The downside is that the same template may be less relevant if evaluated later.
```

Comment:
I think we should have some way to select classes of pods; probably by label.  The default templates should use selectors instead of explicitly-named pods.

Reply:
Let's do this in a follow-up PR.